[Blackjack - Step2] Submit codebrew42's mission

jieun-pak opened the pull request:
Hello! 

This pull request delivers a complete implementation of the **Blackjack game**, including logic for calculating and displaying the final result. The code has been improved for clarity and structure, with some unused test code removed and documentation updated.

---

### Focus

The goal was to **implement only the essential features** needed to meet the deadline, with a focus on minimalism and completeness over perfection.

---

### Challenges & Reflections

* I wasn’t able to fully apply **TDD**, mainly due to time constraints and still getting familiar with Kotlin. I hope to improve this in future projects.
* I made a push to clean up the code on the final day, but I realized some design decisions were flawed from the start, which made refactoring difficult.
* I considered adding **inheritance** for better design, but due to time limits, I prioritized submitting a working version. I plan to refactor separately to experiment with a better structure.

---

### Questions for Reviewers

1. **Companion Object Placement**
   I’m unsure if the following constants really belong in the controller’s companion object. It feels like they should live in a subclass or utility class. Would love your thoughts:

   ```kotlin
   companion object {
       const val DEALER = "dealer"
       const val BLACKJACK_SCORE = 21
   }
   ```

2. **Kotlin Best Practices**
   I’m aware some patterns may not follow idiomatic Kotlin. Could you point out the most un-Kotlin-like parts you noticed, from your perspective?

---

Thanks so much in advance for taking the time to review!
Looking forward to your feedback 🙏

Commit:
docs(README): add 'Card', 'Rank', 'Suit' part of feature list

Commit:
feat: enum class 'Rank'

Commit:
feat: update test cases of 'PlayerCardsTest'

Commit:
style(Card): Remove extra lines

Commit:
add: add algorithm to convert Ace's Rank to 1 if needed

Commit:
add: add symbols to enum class 'Suit'

Commit:
doc(README): add additional classes like 'Person', 'Player' to the feature list

Commit:
feat: add value class `GamblerInfo` to handle user information

Commit:
refactor(PlayerCard): make the values private and expose with modification security

Commit:
add(Player): create a data class

Commit:
add(Deck): functionality of generating and drawing a card/s

Commit:
refactor(PlayerCard): change parameter type so List can be accepted

Commit:
style: change class and variables' name from 'PlayerCards' to 'Hand'

Commit:
style: change class and variables' name from 'PlayerCards' to 'Hand'

Commit:
chore: remove unnecessary import

Commit:
style: incorporate kotlin formatting

Commit:
feat: Added method of toString in data class

Commit:
feat: implemented controller, view

Commit:
feat: add functionality 'display names of players'

Commit:
fix: typo mistake

Commit:
refactor: improve portability and readability

Commit:
chore: add TODO

Commit:
chore: include unnecessary class to the class 'Player'

Commit:
feat: add output function

Commit:
chore: fix formatting

Commit:
feat: add function of displaying players' cards

Commit:
chore: removed unnecessary lines and file

Commit:
feat: add a function of displaying players' cards

Commit:
feat: add a function of taking turns and display its result

Commit:
add: write a helper function in the file extensions

Commit:
add: a function for displaying dealers turn

Commit:
add: a function for displaying final cards with score

Commit:
add: functions for displaying final results

Commit:
add: function for calculation and displaying final results

Commit:
style: add newlines to console

Commit:
refactor: move files to corresponding directories

Commit:
refactor: move winning calculation logic to a new class

Commit:
add: error message in class 'GamblerInfo'

Commit:
add: check boxes which are done, add

jieun-pak committed:
Update src/main/kotlin/blackjack/view/InputView.kt

fix: replace readLine with readlnOrNull

Co-authored-by: Dani <leedaeun.dev@gmail.com>

Commit:
add: add error message which won't be reached

Commit:
add: new class

Commit:
add: helper functions for

Commit:
add(README.md): new features of step2 and updates to step1

Commit:
add: function to check blackjack score with 2 cards

Commit:
add: Betting System Infrastructure

Commit:
refactor: move isHitOrStand from extension.kt to controller

Commit:
add: update calculation of final result

Commit:
docs(READEME.md): separate updates for the future

Commit:
add: calculate and display the new final result

Commit:
chore: fix formatting

Commit:
chore(GamblerInfoTest.kt): remove unnecessary test

Commit:
refactor: replace var with val

Commit:
style: remove abundant `it`

Commit:
doc(README.md): update progress

Commit:
test: remove unused parameters

Commit:
style: improve readability

Commit:
style: replace return if with return when

Commit:
style: remove comments from logic

Commit:
style: enhance readability by modifiying some lines

Commit:
remove(ControllerTest.kt): unnecessary file

Commit:
style: enhance readability

Commit:
style: remove '_' of the val name

Commit:
style: remove empty constructor parentheses

Commit:
style: add magic string as constant

Commit:
fix: change private val to const val for performance and clarity

Commit:
refactor: separate controller's method to different classes

Commit:
style: enhance readability

Commit:
docs(README.md): change the order for better readability and update

Commit:
Merge remote-tracking branch 'upstream/codebrew42' into step2

# Conflicts:
#	README.md
#	src/main/kotlin/blackjack/controller/Controller.kt
#	src/main/kotlin/blackjack/model/Cards.kt
#	src/main/kotlin/blackjack/model/Deck.kt
#	src/main/kotlin/blackjack/model/FinalResult.kt
#	src/main/kotlin/blackjack/model/Player.kt
#	src/main/kotlin/blackjack/view/ErrorPrompt.kt
#	src/main/kotlin/blackjack/view/InputPrompt.kt
#	src/main/kotlin/blackjack/view/InputView.kt
#	src/main/kotlin/blackjack/view/OutputView.kt
#	src/test/kotlin/blackjack/FinalResultTest.kt
#	src/test/kotlin/blackjack/PlayerTest.kt

da-nyee commented on README.md:
```diff
@@ -1,7 +1,37 @@
 
+### Enhanced Blackjack Rules
+- [x] Update `Player.isBlackJack()` to check exactly 2 cards with total score 21
+- [] Player loses entire bet when busting (total > 21)
```
How about checking this box if it looks good?

da-nyee commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -2,115 +2,62 @@ package blackjack.controller
+    private val inputProcessor = InputProcessor()
+    private val gameOrchestrator = GameOrchestrator(Deck(), inputProcessor)
 
     fun run() {
```
<img width="1552" height="987" alt="image" src="https://github.com/user-attachments/assets/878b27e5-e99e-499d-b0c4-fccb148c4d5e" />

You implemented the requirements well, but it looks like `pobi` and `jason` should have earned double 👀

According to the rule:
```
If the dealer busts (exceeds 21),
all remaining players automatically win and receive payouts based on their bets.
```

jieun-pak replied:
Thank you for your feedback!
I'm still a bit unsure about your point, so I wanted to clarify my understanding:
- When the dealer busts and the player has a blackjack: the player receives 1.5× their bet 
-When the dealer busts and the player has a normal hand: the player receives 1× their bet

This seems to follow the standard blackjack payout rules.
The requirement mentions that players receive “payouts based on their bets,” but it doesn’t explicitly say they should receive double (2×).

Please let me know if I misunderstood or missed something!

da-nyee replied:
@codebrew42 

Well, I understood that in that situation, players don’t actually earn double the amount they bet, but rather end up with double their original amount — since the rule says `payouts are based on their bets`.
So I assumed that `pobi` earned 20,000 (10,000 payout + 10,000 bet returned), and `jason` earned 40,000.

da-nyee commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -2,115 +2,62 @@ package blackjack.controller
     companion object {
-        private const val MAX_ATTEMPTS = 5
+        const val DEALER = "dealer"
         const val BLACKJACK_SCORE = 21
```
> **Companion Object Placement**
I’m unsure if the following constants really belong in the controller’s companion object. It feels like they should live in a subclass or utility class. Would love your thoughts:

Well.. I think so, since they're used across multiple classes.
How about introducing an `object` to make things more concise?

da-nyee commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -2,115 +2,62 @@ package blackjack.controller
-    private fun createPlayers() {
-        players =
-            processPlayerNames().map { Player(it) }
+    internal fun createPlayers(): List<Player> {
```
Is there any reason you made this method `internal`?

jieun-pak replied:
I learned about the internal modifier and wanted to try it, thinking it would allow testing while keeping the function hidden from external access. However, I later realized that using internal just for testing can lead to test-driven architecture decisions, something generally considered an anti-pattern. I've since made the method properly public to reflect its actual role and usage within the codebase.

da-nyee commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -2,115 +2,62 @@ package blackjack.controller
-    private fun roundOne() {
-        dealer.addCard(deck.drawCard(2))
-        players.forEach { it.addCard(deck.drawCard(2)) }
+    fun createDealer(): Player {
```
Consider making this method `private` 👀

jieun-pak replied:
I did, thanks!

da-nyee commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -2,115 +2,62 @@ package blackjack.controller
-            }
-        }
-        throw IllegalArgumentException(MAX_ATTEMPT_MESSAGE)
+    private fun roundTwo(
```
I see that you've separated the methods by round.

If one more round were added, how would you handle it?
Would you add a method like `roundThird`?

You might consider refactoring the round-related methods 👀

jieun-pak replied:
That's really good point which we missed, this feedback helped me a lot to think about the proper structure for the next project. I implemented that, thanks again!

```
  Example of adding a new round:
  val gameRounds = listOf(
      ::initialDealRound,
      ::playerActionsRound,
      ::bonusRound,        // I'll add this line later on, if needed.
  )

```

da-nyee commented on src/main/kotlin/blackjack/model/Deck.kt:
```diff
@@ -1,19 +1,19 @@
 
 class Deck {
-    private val _cards = Cards()
+    private val deckCards = Cards()
```
Nice naming 👍

da-nyee commented on src/main/kotlin/blackjack/model/FinalResult.kt:
```diff
@@ -1,22 +1,50 @@
-        private set
-    var draw: List<Player> = listOf()
-        private set
+class FinalResult(val dealer: Player, private val players: List<Player>) {
```
How about wrapping the collection `players`?

da-nyee commented on src/main/kotlin/blackjack/model/Player.kt:
```diff
@@ -28,12 +32,28 @@ class Player(val gamblerInfo: GamblerInfo) {
+        } else if (score > BLACKJACK_SCORE) {
+            status.isBusted = true
+            status.isNeitherBlackjackNorBusted = false
+        }
```
```suggestion
        if (score == BLACKJACK_SCORE && _cards.size == 2) {
            status.apply { 
                isBlackjack = true
                isNeitherBlackjackNorBusted = false
            }
        } else if (score > BLACKJACK_SCORE) {
            status.apply { 
                isBusted = true
                isNeitherBlackjackNorBusted = false
            }
        }
```

You could use `apply` here to make the code more idiomatic Kotlin and improve readability — just a suggestion ☺️

da-nyee commented on src/main/kotlin/blackjack/model/Status.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack.model
+
+data class Status(
```
Have you heard of the State Pattern?
It could be appropriate here — just something to consider 👀

da-nyee commented on src/main/kotlin/blackjack/service/InputProcessor.kt:
```diff
@@ -0,0 +1,82 @@
+import blackjack.view.InputView
+import blackjack.view.OutputView
+
+class InputProcessor {
```
Nice try 👍

da-nyee commented on src/main/kotlin/blackjack/service/GameOrchestrator.kt:
```diff
@@ -0,0 +1,49 @@
+                    OutputView.displayCardsOfPlayers(player)
+                }
+            }
+        }
```
What about extracting the inner condition into a separate method?

Since the rule says:
```
Keep indentation depth to a maximum of 1 (do not exceed 2 levels).
```

da-nyee commented on src/main/kotlin/blackjack/service/GameOrchestrator.kt:
```diff
@@ -0,0 +1,49 @@
+                dealer.addCard(deck.drawCard())
+            }
+        }
+    }
```
This one too 👀

da-nyee commented on src/main/kotlin/blackjack/service/GameOrchestrator.kt:
```diff
@@ -0,0 +1,49 @@
+import blackjack.model.Player
+import blackjack.view.OutputView
+
+class GameOrchestrator(private val deck: Deck, private val inputProcessor: InputProcessor) {
```
How about adding unit tests for this?

da-nyee commented on src/test/kotlin/blackjack/PlayerTest.kt:
```diff
@@ -72,16 +75,117 @@ class PlayerTest {
+            )
+        }
+    }
 }
```
This comment is for `GamblerInfoTest`:

Have you heard of `@ParameterizedTest`?
I think it's a good fit for the tests, so I'd recommend giving it a try 💪

da-nyee reviewed (changes requested):
Hi Jin! 🙌
Great work so far 👍

I've answered your questions and added a few comments — please take a look.
Let me know if anything's unclear!

Commit:
docs(README.md): update progress

Commit:
add(Constants.kt): introduce a new object for handling constants

Commit:
add(Constants.kt): introduce a new object for handling constants

Commit:
remove: unnecessary keyword `internal`

Commit:
refactor(GameOrchestrator.kt): extract methods to reduce indentation depth

Commit:
refactor(GameOrchestrator.kt): modify method to reduce indentation depth

jieun-pak committed:
style(Player.kt): implement `apply` for better readability

Co-authored-by: Dani <leedaeun.dev@gmail.com>

Commit:
Merge branch 'step2' of github.com:codebrew42/kotlin-blackjack into step2

Commit:
refactor: add new class Players to wrap multiple players

Commit:
refactor: add new class Players to wrap multiple players

Commit:
chore: remove whitespaces

Commit:
refactor: improve the round system for better scalability

Commit:
docs: add TODO lists for updating the code

Commit:
add(GamblerInfoTest.kt): add parameterizedTest

Commit:
add(GameOrchestratorTest.kt): add a test case

Commit:
style(Status.kt): change data class to sealed class for better readability

Commit:
docs(README.md): update lists and add things to implement next time

Commit:
style(Status.kt): change data class to sealed class for better readability

da-nyee commented on src/test/kotlin/blackjack/FinalResultTest.kt:
```diff
@@ -2,84 +2,77 @@ package blackjack
-        val dealer = Player(GamblerInfo("dealer"))
-        val player = Player(GamblerInfo("player1"))
+    @Test
+    fun `test blackjack payout scenarios`() {
```
<img width="1552" height="987" alt="image" src="https://github.com/user-attachments/assets/52555e2b-0e25-48a2-a292-fd24d0e6685a" />

Could you check this test?

jieun-pak replied:
I'm really sorry, I know this is basic stuff, but I rushed a bit. I've fixed it now. Thanks so much for your patience!

da-nyee commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -1,116 +1,72 @@
+        dealer: Player,
+        players: List<Player>,
+    ) {
+        val gameRounds =
+            listOf(
+                ::initialDealRound,
+                ::playerActionsRound,
+            )
+        gameRounds.forEach { round -> round(dealer, players) }
```
👍

da-nyee commented on src/main/kotlin/blackjack/model/Players.kt:
```diff
@@ -0,0 +1,15 @@
+package blackjack.model
+
+class Players(private val playerList: List<Player>) {
```
👍

da-nyee commented on src/main/kotlin/blackjack/model/Players.kt:
```diff
@@ -0,0 +1,15 @@
+        playerList.forEach(action)
+    }
+
+    fun calculateTotalEarnings(): Double {
```
It seems like this method is never used 👀

jieun-pak replied:
Oops, sorry. I’ve removed it. Thank you!

da-nyee commented on src/main/kotlin/blackjack/model/Status.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack.model
+
+sealed class Status {
```
You used `sealed class` 👍

da-nyee commented on src/main/kotlin/blackjack/model/Status.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack.model
+
+sealed class Status {
+    object Normal : Status()
+    object Blackjack : Status()
+    object Busted : Status()
```
👍

da-nyee commented on src/test/kotlin/blackjack/GamblerInfoTest.kt:
```diff
@@ -1,27 +1,31 @@
 class GamblerInfoTest {
-    @Test
-    fun `should throw exception - empty input`() {
+    @ParameterizedTest
+    @MethodSource("invalidNames")
```
You made it 👍

da-nyee reviewed (changes requested):
Hi Jin! 🙌

Thanks for applying the feedback — I've gone through all the changes.
Everything looks good, but there's just one test failure — could you take another look?
I've also left a few comments, so feel free to check them out ☺️

Commit:
remove: unused method

Commit:
style: add newlines between objects

Commit:
fix: fix the wrong payout logic

The payout logic was checking dealer's status is busted before player's status is blackjack, so players with blackjack against a busted dealer only got normal win rate (1.0x) instead of blackjack rate (1.5x)

Commit:
chore: change setting

jieun-pak commented:
I incorporated all of your feedback, and learned a lot from you. Thank you so much !

By the way, I was wondering if you have a preferred way to detect unused functions.
I know IntelliJ marks them in gray, but I was curious if there's a more reliable or systematic method you personally like to use. 🙂

Commit:
style: controller uses a single deck instance to avoid potential issues

Commit:
fix: add one line according to the requirement

Commit:
style: improve naming of variables

Commit:
style: extract magic number to the companion object

Commit:
fix: add newlines according to the requirement

da-nyee commented on src/test/kotlin/blackjack/FinalResultTest.kt:
```diff
@@ -2,84 +2,77 @@ package blackjack
-import java.util.stream.Stream
+import org.junit.jupiter.api.Test
 
 class FinalResultTest {
```
<img width="570" height="110" alt="image" src="https://github.com/user-attachments/assets/ad625777-110f-43c1-87cb-bdda4ef9e7a5" />

All good 👍

da-nyee commented on src/test/kotlin/blackjack/PlayerTest.kt:
```diff
@@ -72,16 +76,119 @@ class PlayerTest {
+            )
+        }
+    }
 }
```
> By the way, I was wondering if you have a preferred way to detect unused functions.
I know IntelliJ marks them in gray, but I was curious if there's a more reliable or systematic method you personally like to use. 🙂

I don't have a specific way to check for unused methods either 😅
I usually rely on IntelliJ, just like you 👀

jieun-pak replied:
I see, thank you ! :)

da-nyee reviewed (approved):
Hi Jin! 🙌

It might have been a tough one, but you did a great job 💯
Everything looks solid, so I think it's a good time to wrap up the mission!
I really hope the feedback was helpful and that you'll stay interested in development.

If you have any further questions, feel free to reach out — I'll be happy to help ☺️
